overlord/configstate/configcore/certs.go

```diff
@@ -23,15 +24,241 @@ package configcore
+		return nil, fmt.Errorf("cannot read certs directory: %v", err)
+	}
+
+	var certsObjects []certificate
+	for _, cert := range certFiles {
+		if cert.IsDir() {
+			continue
+		}
+
+		// So what are we trying to achieve here? When provided with certificate directories
+		// they may or may not be symbolic links to the actual certificate file. To handle this
+		// correctly we evaluate the symbolic links and get the real path of the certificate, because
+		// we are not going to 'copy' or link to the symbolic link, instead we will recreate the symbolic
+		// link to the real path of the certificate.
+		certRealPath := filepath.Join(certsPath, cert.Name())
+		if cert.Mode()&os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(certRealPath)
+			if err != nil {
+				return nil, fmt.Errorf("cannot resolve symbolic link: %v", err)
+			}
+			certRealPath = resolvedPath
+		}
+
+		// If the file is not a symbolic link then Path and RealPath will be identical.
+		certObject := certificate{
+			Name:     cert.Name(),
+			Path:     filepath.Join(certsPath, cert.Name()),
+			RealPath: certRealPath,
+		}
+		certsObjects = append(certsObjects, certObject)
+	}
+	return certsObjects, nil
+}
+
+// filterCerts filters out the certificates that are not allowed to be merged.
+func filterCerts(certs []certificate, blockedCerts []string) ([]certificate, error) {
+	var filteredCerts []certificate
+	for _, cert := range certs {
+		if isBlocked(cert, blockedCerts) {
+			continue
+		}
+		filteredCerts = append(filteredCerts, cert)
+	}
+	return filteredCerts, nil
+}
+
+func filterCertsInDirectory(dirPath string, blockedCerts []string) error {
+	existingCerts, err := getCertObjects(dirPath)
+	if err != nil {
+		return err
+	}
+
+	// Remove the certs that are marked blocked.
+	for _, cert := range existingCerts {
+		if isBlocked(cert, blockedCerts) {
+			if err := os.Remove(cert.Path); err != nil {
+				return fmt.Errorf("cannot remove blocked certificate: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// installCerts populates symbolic links in the output directory, to each certificate
+// provided. There may need to be some provide apparmor rules for the source directories
+// and not just the merged (output) directory.
+// TODO apparmor rules for the source directories?
+func installCerts(outputPath string, certs []certificate) error {
+	for _, cert := range certs {
+		// When updating existing certifications we want to make sure we don't
+		// overwrite the certification with itself.
+		certDestinationPath := filepath.Join(outputPath, cert.Name)
+		if certDestinationPath == cert.RealPath {
+			continue
+		}
+
+		// If the path exists, then remove it first to update the certificate
+		_, err := os.Lstat(certDestinationPath)
+		if err == nil {
+			err := os.Remove(certDestinationPath)
+			if err != nil {
+				return fmt.Errorf("cannot remove existing certificate: %v", err)
+			}
+		} else if e, ok := err.(*os.PathError); !ok || e.Err != syscall.ENOENT {
```

Comment:
This is equivalent and IMO a bit simpler
```suggestion
		} else !errors.Is(err, os.ErrNotExist) {
```